keymap.c

```diff
@@ -1298,6 +1298,186 @@ enum CommandResult mutt_parse_bind(struct Buffer *buf, struct Buffer *s,
+    }
+    cur = next;
+  }
+
+  if (last != NULL)
+    last->next = NULL;
+
+  *map = first;
+}
+
+/**
+ * mutt_parse_unbind - Parse the 'unbind' command - Implements ::command_t
+ * @param buf
+ * @param s
+ * @param data
+ * @param err
+ *
+ * @return r
+ *
+ * Command unbinds specific binging in menu-name, in all menu-names
+ * or all bindings in all menu-names.
+ * unbind <menu-name[,…]|*> [<key_sequence>]
+ */
+enum CommandResult mutt_parse_unbind(struct Buffer *buf, struct Buffer *s,
+                                     unsigned long data, struct Buffer *err)
+{
+  int menu[sizeof(Menus) / sizeof(struct Mapping) - 1], nummenus;
+  enum CommandResult r = MUTT_CMD_SUCCESS;
+  bool all_menus = false, all_keys = false;
+  char *key = NULL;
+  char *cmd = "unbind";
+  if ((data & MUTT_UNMACRO) == MUTT_UNMACRO)
+    cmd = "unmacro";
+
+  mutt_extract_token(buf, s, 0);
+  if (mutt_str_strcmp(buf->data, "*") == 0) //when menu-names is * (all)
+    all_menus = true;
+  else // when menu-names is comma separated list
+    parse_menu(menu, buf->data, mutt_array_size(menu), &nummenus, err, cmd);
+
+  if (MoreArgs(s)) // second arg - key
+  {
+    mutt_extract_token(buf, s, 0);
+    key = buf->data;
+  }
+  else
+    all_keys = true;
+
+  if ((all_menus && mutt_str_strcmp(key, "q") == 0) ||
+      (all_menus && mutt_str_strcmp(key, ":") == 0))
+  {
+    mutt_buffer_printf(err, _("%s: cannot unbind '%s' from all menus"), cmd, key);
+    return MUTT_CMD_ERROR;
+  }
+
+  if (MoreArgs(s))
+  {
+    mutt_buffer_printf(err, _("%s: too many arguments"), cmd);
+    return MUTT_CMD_ERROR;
+  }
+  else // Here comes the logic, Mr. Tuvok.
+  {
+    if (all_menus == true && all_keys == true)
+    {
+      for (int i = 0; i < MENU_MAX; ++i)
+      {
+        km_unbind_all(&Keymaps[i],data);
+        if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
+        {
+          km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
+          km_bindkey("q", MENU_GENERIC, OP_QUIT);
+          km_bindkey("?", MENU_GENERIC, OP_HELP);
+        }
+      }
+    }
+    else if (all_menus == false && all_keys == true)
+    {
+      for (int i = 0; i < nummenus; ++i)
+      {
+        km_unbind_all(&Keymaps[menu[i]],data);
+        if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
+        {
+          km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
+          km_bindkey("q", MENU_GENERIC, OP_QUIT);
+          km_bindkey("?", MENU_GENERIC, OP_HELP);
+        }
+      }
+    }
+    else if (all_menus == true && all_keys == false)
+    {
+      for (int i = 0; i < MENU_MAX; ++i)
+        km_bindkey(key, i, OP_NULL);
+    }
+    else // unbind key from specified menus
```

Comment:
Similarly, this can be merged into:

```C
else {
    int upper_bound = all_menus ? MENU_MAX : nummenus;
    for (int i = 0; i < upper_bound; ++i) {
        target_menu = all_menus ? menu[i] : i;
        km_bindkey(key, target_menu, OP_NULL);
    }
}
```

Reply:
this segfaults in km_bindkey(). I'll take a look at it tomorrow.

Reply:
@jindraj that is my fault. i made a mistake w/ `target_menu =...`. correct line is `target_menu = all_menus ? i : menu[i];`

I'll triple check my messages now to save you any headache  :)